### ISSUE TITLE
Add Radxa Taco - the successor of the ROCK Pi SATA HAT(Penta SATA CM4 carrier board)

### ISSUE THREAD

**Issue by hipboi** [geerlingguy/raspberry-pi-pcie-devices#202]
I've been following this repository for quite long time. As a Raspberry Pi fan, we've been working on the Radxa Taco for a while, which is an ultimate CM4 carrier board designed for nas/server/router applications. The ideas is based on the Quad SATA HAT for the Raspberry Pi 4 and go further beyond it. The Radxa Taco has the following features:

- Use asm1806 pcie switch
- Dual ethernet, one 1000M from CM4, one 2.5G rtk8125b to pcie switch
- 5x SATA via JMB585 to pcie switch
- 1x M.2 M key NVMe to pcie switch
- 1x PCIe M.2 WiFi module, support Intel AX200
- 10P header for Top Board with OLED and button
- RTC/USB 2/uSD card

The case is working on...

![radxa_taco_top](https://user-images.githubusercontent.com/693050/130958517-967f11d1-643c-433b-b873-e343b4896ce8.jpg)
![radxa_taco_angle](https://user-images.githubusercontent.com/693050/130958539-69dfb4b9-c595-4f7e-9851-a1844efb6fc4.jpg)
![radxa_taco_bottom](https://user-images.githubusercontent.com/693050/130958555-e8e284be-6343-4ffc-b2bd-50e250548237.jpg)

**Comment by tsaotse** [comment]
When will it be released?

**Comment by geerlingguy** [comment]
I _just_ received my board today. Like, a couple minutes ago. This thing is very, very neat. Going to be playing around with it soon!

**Comment by andyattebery** [comment]
I spent some time testing the board over the weekend. Here are my initial thoughts and notes:

- The SATA connectors are female 22 pin (7+15) which are great if you are using 2.5" drives, but not so much for 3.5" drives. I bought some [male to male adapters](https://www.amazon.com/gp/product/B00S6HTVGI), so I could plug in standard female 7 pin SATA cables.
- I only had a 3.5A power supply readily available instead of the 5A that was suggested, so I powered the 4 3.5" drives I'm using from a spare computer PSU. Once I get a 5.5mm x 2.5mm barrel adapter for my 8A power supply, I'll test powering the drives directly. However, I didn't see any female to female 15 pin SATA power cables on Amazon, so this might involve hacking together some from spare module PSU cables. (Please tell me if there is a better way.)

![Testing "Setup"](https://user-images.githubusercontent.com/199135/134104165-41b15686-f2f1-4638-91f4-75da1558d6cf.jpeg)

- Compute Modules with eMMC have the micro SD card slot disabled since the SD connectivity goes to the eMMC. Because the board doesn't have a "periphrial" USB port to flash the eMMC, there isn't a great way to directly install an OS to the eMMC without also having an [IO Board](https://www.amazon.com/Waveshare-Raspberry-Development-Reference-Base-Board/dp/B08PVLH3PZ) or similar. I used a USB drive as a boot drive plugged into the USB 2.0 port; surprisingly this wasn't as slow as I was expecting. I suppose once booted from the USB drive, the OS installation could be copied to the eMMC.
- I updated the bootloader to the newest version to support NVMe drive booting as outlined [here](https://github.com/raspberrypi/documentation/blob/develop/documentation/asciidoc/computers/raspberry-pi/boot-nvme.adoc) and by [Jeff](https://www.jeffgeerling.com/blog/2021/raspberry-pi-can-boot-nvme-ssds-now). To avoid needing to boot the full version of Raspberry Pi OS to use the SD Card Copier utility, I used a NVMe to USB adapter to directly install the OS to the drive and initially booted it from USB to install the updates. I discovered that even the newest [bootloader won't boot a NVMe drive that is connected via a PCIe switch](https://www.raspberrypi.org/forums/viewtopic.php?f=29&t=319228#p1911031). This led me to copying the boot partition to a USB drive and leaving the root partition on the NVMe drive; the same kind of setup would work with a boot partition on the eMMC or SD card too. I was able to get this setup to work with Raspberry Pi OS (I only tested the newest 64-bit version)and Ubuntu 21.04.
- Even though the ['proposed' Ubuntu 21.04 kernel](https://launchpad.net/ubuntu/+source/linux-raspi/5.11.0-1018.19) says it enabled SATA drive support, the kernel module didn't load for me. It also looks like the kernel module for the RTL8125 2.5GbE controller wasn't loaded either.
- I haven't tested the 2.5GbE controller on Raspberry Pi OS, but it looks like the kernel driver/module isn't loaded, and I don't see the network interface.

![Raspberry Pi OS 2.5GbE controller](https://user-images.githubusercontent.com/199135/134104306-753708f1-86d6-4172-8d9f-86fb42d7e35c.png)

- I really wanted to test ZFS which is why I was wanting to use Ubuntu since it has the kernel module built-in; however, now I realize I didn't verify the ARM/Raspberry Pi port has it, so maybe that was all for naught. I think my next steps will be [compiling the Raspberry Pi OS kernel](https://redshirtjeff.com/listing/linux-recompile-shirt?product=211) with ZFS support using the [45 Drives repo](https://github.com/45Drives/linux) they used for the "Mini Storinator".
- Before I attempt to get ZFS working, I'll try to get some benchmarks for various `mdadm` RAID configurations.

**Comment by geerlingguy** [comment]
Closing this issue, and let's focus on adding test data to the related issue: #268.

**Comment by rejsmont** [comment]
Hi there!

I am running a 40TB storage cluster using 2 RaspberryPIs 4s with 8GB RAM with two 5xSATA-to-USB3 enclosures, each with 5x4TB drives. Currently under FreeBSD with ZFS in raidz mode with the two nodes glued together using GlusterFS, with an additional RaspberryPI 4 with 4GB RAM as an arbiter node. Would love to compare the performance to CM4s and this board!

Any chance to buy these soon? Are you interested in ZFS+GlusterFS performance figures?

Cheers,

Radek

**Comment by geerlingguy** [comment]
@rejsmont - ZFS + GlusterFS on top of two Raspberry Pi 4s with USB-to-SATA adapters, you _do_ live on the edge!

**Comment by rejsmont** [comment]
> @rejsmont - ZFS + GlusterFS on top of two Raspberry Pi 4s with USB-to-SATA adapters, you _do_ live on the edge!

Knock on the wood - uptime of 3 months now, mean load of 20GB/day, zero issues thus far... I guess credit goes to FreeBSD usb-storage driver implementation 😅

**Comment by WaaromZoMoeilijk** [comment]
> > @rejsmont - ZFS + GlusterFS on top of two Raspberry Pi 4s with USB-to-SATA adapters, you _do_ live on the edge!
> 
> Knock on the wood - uptime of 3 months now, mean load of 20GB/day, zero issues thus far... I guess credit goes to the FreeBSD usb-storage driver implementation 😅

Yes I want to know more!!!! This is really bleeding edge 😂 could you share a little more? Can we DM on GitHub btw?

**Comment by ishi** [comment]
Is it possible by any chance to receive one of these for testing. I would love to test it. Perfect fit for home NAS with some additional compute capabilities 🤩

**Comment by vebmaster** [comment]
Hello.
The page https://wiki.radxa.com/Taco says:
One M.2 M Key for NVMe SSD
One M.2 E Key for WiFi 6 modules or Google Coral AI module
But in the photo there is only one M.2 port. Where is the second one?

Where can you buy this board?

**Comment by hipboi** [comment]
@vebmaster 

The new hw revision changes to two m.2. This board is only available in small sample quantity at the moment.

**Comment by vebmaster** [comment]
Radxa Taco only supports "Raspberry Pi CM4" or can I connect others? If others are possible, which ones?

If you use the Coral Edge module, then the SOM "Raspberry Pi CM4" will have problems:
"Installation failing on Raspberry Pi CM4 for PCI-E driver" - https://github.com/google-coral/edgetpu/issues/280

**Comment by hipboi** [comment]
The Radxa CM3 RK3566 PCIe support:

- Support MSI with Per-Vector Masking (PVM) and 32 multiple MSI
- Support MSI-X with Per-Vector Masking (PVM) and 64 MSI-X table size

**Comment by vebmaster** [comment]
> This board is only available in small sample quantity at the moment.

Is it possible to buy 1 board from you (with two m.2 slots)? If so, how much will it cost?

**Comment by WhiteRose96** [comment]
> 

Hello, I've been following this project for quite some time. Very excited to see the final product. I'm interested in getting my hands on a sample if they are still available. Would absolutely love to help test.

**Comment by WhiteRose96** [comment]
> 



> @vebmaster
> 
> The new hw revision changes to two m.2. This board is only available in small sample quantity at the moment.

Hello, I've been following this project for quite some time. Very excited to see the final product. I'm interested in getting my hands on a sample if they are still available. Would absolutely love to help test.

**Comment by geerlingguy** [comment]
Please don't use this issue for product status information; please contact Radxa directly for that, or hop into their Discord to discuss availability. There aren't any more samples AFAICT, and they still plan on releasing the board for sale sometime before the end of this year.

**Comment by hipboi** [comment]
Anyone who is interested in the Taco board, can subscribe the notification here:

https://shop.allnetchina.cn/collections/sata-hat/products/taco

**Comment by ogg1e** [comment]
Any updates on case for 3.5 drives? I'd be happy with stl files to print on a 3d printer.

**Comment by mrbrazzi** [comment]
Hey everyone, I recently got my hands on the Raspberry Pi Compute Module 4 with 8GB RAM and 32GB EMMC, along with the Raxda Taco board. I also picked up an AC Adapter 12V 5.0A and a USB 2.0 cable A-A M-M.

I've been trying to get my CM4 to boot, but despite my extensive research and various troubleshooting steps, I haven't been successful.

Here's what I've tried so far:
1. Connected the board to the AC Adapter. The red power LED is on and steady, and the 4 Activity LEDs blink once.
2. Tried pressing and holding the Maskrom Key while connecting the board to the USB 2.0 cable and my Dell Latitude 5531 laptop with Windows 11. The red power LED stays on, the 4 Activity LEDs blink once, and there's an additional steady but low-intensity blue LED near the Maskrom Key. After running rpiboot, it just stays waiting.
3. Connected the board to the AC Adapter and the laptop using the USB 2.0 cable. The red power LED is on and steady, and the 4 Activity LEDs blink once.

Unfortunately, my Dell monitor isn't receiving any HDMI signal in all cases.

In the attached images, you will notice the red power LED and the blue LED.

@hipboi I would appreciate any ideas or suggestions that might help me boot my Raxda Taco with the CM4. 

Thank you very much in advance.

![1722639026579](https://github.com/user-attachments/assets/f2d8865b-46c8-4dc9-b8f5-55f25274978c)
![1722639026586](https://github.com/user-attachments/assets/aebf2fc8-539e-4d2f-9376-ccbddfe56218)
![1722639026593](https://github.com/user-attachments/assets/c85929a6-4e87-4fa1-95d2-f347e1e3b331)
![1722639026603](https://github.com/user-attachments/assets/3c8ebdb5-fdd4-442c-a04a-a2e14a995764)
